android-pdf-viewer/src/main/java/com/github/barteksc/pdfviewer/annotation/core/pdf/PdfUtil.kt

```diff
@@ -0,0 +1,367 @@
+            val annots: PdfArray? = page.getAsArray(PdfName.ANNOTS)
+            if (annots == null) {
+                logDebug(TAG, "Annotations array for page $i is null")
+
+            } else {
+                logDebug(TAG, "Annotations array for page $i: $annots")
+
+                for (j in 0 until annots.size()) {
+                    val annotation: PdfDictionary = annots.getAsDict(j)
+
+                    // Extract extras
+                    val entityLinks: PdfArray? = annotation.getAsArray(PdfName("EntityLinks"))
+                    if (entityLinks != null) {
+                        for (k in 0 until entityLinks.size()) {
+                            val entityLink: PdfDictionary = entityLinks.getAsDict(k)
+                            val relKey: PdfString? = entityLink.getAsString(PdfName("relKey"))
+                            val relType: PdfString? = entityLink.getAsString(PdfName("relType"))
+
+                            logDebug(TAG, "Annotation $j on page $i - EntityLink $k:")
+                            logDebug(TAG, "  relKey: $relKey")
+                            logDebug(TAG, "  relType: $relType")
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /** Uses the passed PDF file to create a PNG image from the first page,
+     *  maps the PDF annotations to shapes that will be saved as json string */
+    @JvmStatic
+    fun getPdfToImageResultData(
+        pdfFilePath: String,
+        outputDirectory: String
+    ): PdfToImageResultData {
+        // (PDF, annotations) -> (PNG, shapes)
+        val resultData = convertPdfAnnotationsToPngShapes(pdfFilePath, outputDirectory)
+        // Remove annotations from the PDF so we can draw shapes from MeasureLib on the same PDF
+        AnnotationManager.removeAnnotationsFromPdf(pdfFilePath)
+        return resultData
+    }
+
+
+    /** Maps shapes to annotations and draws them to the given PDF */
+    @JvmStatic
+    fun getResultPdf(pdfFile: File, pdfPageHeight: Int, jsonShapes: String) {
+        val shapes = mapJsonStringToPdfShapes(jsonShapes)
+        drawPngShapesToPdf(pdfFile, pdfPageHeight, shapes)
+    }
+
+    /** Extract the annotations for the given PDF file path and page number.
+     * Page number is always 1 for now */
+    @Throws(IOException::class)
+    @JvmStatic
+    fun getAnnotationsFrom(filePath: String, pageNum: Int): List<Annotation> {
+        if (filePath.isEmpty()) throw Exception("Input file is empty")
+        val file = File(filePath)
+        if (!file.exists()) throw Exception("Input file does not exist")
+
+        // a list of the annotations
+        val annotationsList = mutableListOf<Annotation>()
+
+        // input stream from file
+        val inputStream: InputStream = FileInputStream(file)
+
+        // we create a reader for a certain document
+        val reader = PdfReader(inputStream)
+
+        // read annotations for the given page
```

Comment:
Remove white lines and redundant comments - make sure to apply this to the whole file